File: docs/legacy(old PACE)/unsupported/VB.NET/PaceWebServiceSDK/util/AttachmentService.vb
Imports PaceWebServiceSDK.efipaceservices
Imports System.IO

Public Class AttachmentService

    Public Shared Sub Run()
        Console.WriteLine("Enter the file name to attach")
        Dim attachmentFileName As String = Console.ReadLine()

        Console.WriteLine("Enter the mime type")
        Dim mimeType As String = Console.ReadLine()

        If Not File.Exists(attachmentFileName) Then
            Console.WriteLine("{0} does not exist!", attachmentFileName)
            Return
        End If

        Console.WriteLine("Processing...")

        Dim content As Byte() = readFile(attachmentFileName)
        Dim fileNameParts As [String]() = attachmentFileName.Split("."c).ToArray()
        Dim fileName As [String] = fileNameParts(0)
        Dim extension As [String] = If((fileNameParts.Length > 1), fileNameParts(1), "")

        Dim attachment As Attachment = New Attachment()
        attachment.content = Convert.ToBase64String(content)
        attachment.fileExtension = extension
        attachment.mimeType = mimeType
        attachment.name = fileName


        Dim attachmentServiceHttpBinding As AttachmentServiceHttpBinding = PaceClient.getAttachmentServiceHttpBinding()

        '1. Add Attachment
        Console.WriteLine("Enter the Base Object")
        Dim baseObject As String = Console.ReadLine()

        Console.WriteLine("Enter the Primary key")
        Dim primaryKey As String = Console.ReadLine()

        Console.WriteLine("Enter the Attribute")
        Dim attribute As String = Console.ReadLine()

        Dim newKey As String = attachmentServiceHttpBinding.addAttachment(baseObject, primaryKey, attribute, attachment)

        '2. Retrieve the attachment
        Dim attachmentRetrieved As Attachment = attachmentServiceHttpBinding.getAttachmentFromKey(newKey)
        Console.WriteLine(attachmentRetrieved.content)

        '3. Remove all attachments
        Console.WriteLine("Removing all the attachments")
        attachmentServiceHttpBinding.removeAllAttachments(baseObject, primaryKey, attribute)

    End Sub

    Private Shared Function readFile(ByVal attachmentFileName As [String]) As Byte()
        Dim bytes As Byte() = New Byte(-1) {}
        Using fileStream As New FileStream(attachmentFileName, FileMode.Open, FileAccess.Read, FileShare.Read)
            bytes = New Byte(fileStream.Length - 1) {}
            Dim numBytesToRead As Integer = CInt(fileStream.Length)
            Dim numBytesRead As Integer = 0
            While numBytesToRead > 0
                Dim n As Integer = fileStream.Read(bytes, numBytesRead, numBytesToRead)

                If n = 0 Then
                    Exit While
                End If

                numBytesRead += n
                numBytesToRead -= n
            End While
        End Using
        Return bytes
    End Function


End Class

File: docs/legacy(old PACE)/C#/4.0/PaceAPISample/util/AttachmentService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using efipaceservices;
using System.IO;

namespace Pace_Web_Service_SDK.util
{
    class AttachmentService
    {
        private String m_objectName;
        private String m_primaryKey;
        private String m_attribute;
        private Attachment m_attachment;
        public AttachmentService(String objectName, String primaryKey, String attachmentAttributeName, Attachment attachment)
        {
            m_objectName = objectName;
            m_primaryKey = primaryKey;
            m_attribute = attachmentAttributeName;
            m_attachment = attachment;
        }

        public static void Main()
        {
            Console.WriteLine("Enter the mime type");
            String mimeType = Console.ReadLine();

            // Read the attachment
            Console.WriteLine("Enter the file name to attach");
            String attachmentFileName = Console.ReadLine();
            if (!File.Exists(attachmentFileName))
            {
                Console.WriteLine("{0} does not exist!", attachmentFileName);
                return;
            }

            byte[] content = readFile(attachmentFileName);
            String[] fileNameParts = attachmentFileName.Split('.').ToArray();

            String fileName = fileNameParts[0];
            String extension = (fileNameParts.Length > 1) ? fileNameParts[1] : "";

            Attachment attachment = new Attachment();
            attachment.content = Convert.ToBase64String(content);
            attachment.fileExtension = extension;
            attachment.mimeType = mimeType;
            attachment.name = fileName;

            Console.WriteLine("Trying to create a Job");
            Job job = CreateJob.Run();

            Console.WriteLine("Job created");

            Attachment attachmentFromService = AttachmentService.Run("Job", "attachment", job.job, attachment);

            ReadObjectHttpBinding readObjectHttpBinding = PaceClient.getReadObjectHttpBinding();

            Job retrievedJob = readObjectHttpBinding.readJob(job);
            Console.WriteLine("Job " + retrievedJob.job + " has attachment " + retrievedJob.attachment);

            UpdateObjectHttpBinding updateObjectHttpBinding = PaceClient.getUpdateObjectHttpBinding();
            retrievedJob.attachment = null;
            updateObjectHttpBinding.updateJob(retrievedJob);

            Job reretrievedJob = readObjectHttpBinding.readJob(job);
            Console.WriteLine("Job " + reretrievedJob.job + " has attachment " + reretrievedJob.attachment);
        }


        public static Attachment Run(String objectName, String attribute, String primaryKey, Attachment attachment)
        {
            AttachmentServiceHttpBinding attachmentServiceHttpBinding = PaceClient.getAttachmentServiceHttpBinding();

            // trying to set attachment
            Console.WriteLine("Trying to set attachment on object=" + objectName + "." + attribute + ", key=" + primaryKey + ", attachment=" + attachment);

            String newKey = attachmentServiceHttpBinding.setAttachment(objectName, primaryKey, attribute, attachment);

            Console.WriteLine("Saved attachment. The new attachment key = " + newKey);

            // retrieve this attachment back
            Attachment attachmentFromService = attachmentServiceHttpBinding.getAttachmentFromKey(newKey);
            Console.WriteLine("Retrieved the attachment back using key " + newKey);
            Console.WriteLine("Trying to retrieve attachment using the getAttachment method");
            Attachment attachmentFromService2 = attachmentServiceHttpBinding.getAttachment(objectName, primaryKey, attribute);
            Console.WriteLine("Retrieved attachment " + attachmentFromService2);

            return attachmentFromService;
        }

        private static byte[] readFile(String attachmentFileName)
        {
            byte[] bytes = new byte[0];
            using (FileStream fileStream = new FileStream(attachmentFileName, FileMode.Open, FileAccess.Read, FileShare.Read))
            {
                bytes = new byte[fileStream.Length];
                int numBytesToRead = (int)fileStream.Length;
                int numBytesRead = 0;
                while (numBytesToRead > 0)
                {
                    int n = fileStream.Read(bytes, numBytesRead, numBytesToRead);

                    if (n == 0)
                        break;

                    numBytesRead += n;
                    numBytesToRead -= n;
                }
            }
            return bytes;
        }
    }
}
